bspump/kafka/source.py

```diff
@@ -174,9 +238,76 @@ async def main(self):
 						"_kafka_offset": m.offset(),
 					})
 
+					# Store the offset associated with msg to a local cache.
+					# Stored offsets are committed to Kafka by a background thread every 'auto.commit.interval.ms'.
+					# Explicitly storing offsets after processing gives at-least once semantics.
+					try:
+
+						if consumer:
+							consumer.store_offsets(m)
+
+					except confluent_kafka.KafkaException as err:
+
+						# Reballacing of partitions happened during consuming,
+						# so this consumer no longer owns the partition
+						# -> let the other consumer consume the events and here just finish
+						# https://medium.com/@a.a.halutin/simple-examples-with-confluent-kafka-9b7e58534a88
+						if err.args[0].code() == confluent_kafka.KafkaError._STATE:
+							break
+
+						L.warning("The following warning occurred inside Kafka consumer: '{}'".format(err))
+						break
+
+					except RuntimeError as e:
+						L.exception("Error storing offsets, possible consumer state issue: '{}'".format(e))
+						break
+
+				# Manually commit the offsets after processing the batch
+				try:
+
+					if consumer:
+						consumer.commit(asynchronous=False)  # Commit offsets synchronously
+
+				except confluent_kafka.KafkaException as err:
+
+					if err.args[0].code() == confluent_kafka.KafkaError._NO_OFFSET:
+						return
+
+					L.exception("Failed to commit offsets: '{}'".format(err))
+
+
+	async def main(self):
+
+		while self.Running:
+
+			try:
+				c = confluent_kafka.Consumer(self.ConsumerConfig, logger=L)
+
+			except BaseException as e:
+				L.exception("Error when connecting to Kafka")
+				self.Pipeline.set_error(None, None, e)
+				return
+
+			c.subscribe(self.Subscribe)
+
+			try:
+				# Run the poll loop on a separate thread
+				await self.ProactorService.execute(self.poll_kafka, c)
```

Comment:
Don't use ProactorService for this, launch your own thread.

Also consider (for performance reasons) launching more than one consumer thread - in that case, `self.Buffer` concept will have to be extended.